pkg/asset/quota/openstack/openstack.go

```diff
@@ -0,0 +1,96 @@
+	machineapi "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
+	"github.com/openshift/installer/pkg/asset/installconfig/openstack/validation"
+	"github.com/openshift/installer/pkg/quota"
+	"github.com/openshift/installer/pkg/types"
+	openstackprovider "sigs.k8s.io/cluster-api-provider-openstack/pkg/apis/openstackproviderconfig/v1alpha1"
+)
+
+// Constraints returns a list of quota constraints based on the InstallConfig.
+// These constraints can be used to check if there is enough quota for creating a cluster
+// for the install config.
+func Constraints(config *types.InstallConfig, controlPlanes []machineapi.Machine, computes []machineapi.MachineSet) []quota.Constraint {
+	ctrplConfigs := make([]*openstackprovider.OpenstackProviderSpec, len(controlPlanes))
+	for i, m := range controlPlanes {
+		ctrplConfigs[i] = m.Spec.ProviderSpec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+	}
+	computeReplicas := make([]int64, len(computes))
+	computeConfigs := make([]*openstackprovider.OpenstackProviderSpec, len(computes))
+	for i, w := range computes {
+		computeReplicas[i] = int64(*w.Spec.Replicas)
+		computeConfigs[i] = w.Spec.Template.Spec.ProviderSpec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+	}
+
+	ret := controlPlane(config, ctrplConfigs)
+	ret = append(ret, compute(config, computeReplicas, computeConfigs)...)
+	return aggregate(ret)
+}
+
+func controlPlane(config *types.InstallConfig, machines []*openstackprovider.OpenstackProviderSpec) []quota.Constraint {
+	var ret []quota.Constraint
+	for _, m := range machines {
+		ret = append(ret, machineFlavorCoresToQuota(m.Flavor, config), machineFlavorRAMToQuota(m.Flavor, config))
+	}
+	ret = append(ret, instanceConstraint(int64(len(machines))))
+	return ret
+}
+
+func compute(config *types.InstallConfig, replicas []int64, machines []*openstackprovider.OpenstackProviderSpec) []quota.Constraint {
+	var ret []quota.Constraint
+	for idx, m := range machines {
+		coresConstraint := machineFlavorCoresToQuota(m.Flavor, config)
+		coresConstraint.Count = coresConstraint.Count * replicas[idx]
+		ramConstraint := machineFlavorRAMToQuota(m.Flavor, config)
+		ramConstraint.Count = ramConstraint.Count * replicas[idx]
+		ret = append(ret, instanceConstraint(int64(replicas[idx])), coresConstraint, ramConstraint)
+	}
+	return ret
+}
+
+func aggregate(quotas []quota.Constraint) []quota.Constraint {
+	counts := map[string]int64{}
+	for _, q := range quotas {
+		counts[q.Name] = counts[q.Name] + q.Count
+	}
+	aggregatedQuotas := make([]quota.Constraint, 0, len(counts))
+	for n, c := range counts {
+		aggregatedQuotas = append(aggregatedQuotas, quota.Constraint{Name: n, Count: c})
+	}
+	return aggregatedQuotas
+}
+
+//constraintGenerator generates a list of constraints.
+type constraintGenerator func() []quota.Constraint
+
+func machineFlavorCoresToQuota(f string, config *types.InstallConfig) quota.Constraint {
+	var ret quota.Constraint
+	ci, err := validation.GetCloudInfo(config)
```

Comment:
The `GetCloudInfo` function is quite expensive and you call it twice for each machine. How about passing a reference to the `CloudInfo` object to the `machineFlavorCoresToQuota` and `machineFlavorRAMToQuota` functions instead?

Reply:
yeah I did it and the Platform Quota Check is around 3 times faster from my initial testing, thanks!